### PR TITLE
Add `--scrape-tests` option to rustdoc to scrape functions marked `#[test]`

### PR DESCRIPTION
**Pull request by willcrichton**
As a part of stabilizing the scrape examples extension in Cargo, I uncovered a bug where examples cannot be scraped from tests. See this test: https://github.com/rust-lang/cargo/pull/10343/files#diff-27aa4f012ebfebaaee61498d91d2370de460628405d136b05e77efe61e044679R2496

The issue is that when rustdoc is run on a test file, because `--test` is not passed as a rustc option, then functions annotated with `#[test]` are ignored by the compiler. So this PR changes rustdoc so when `--test` is passed in conjunction with a `--scrape-example-<suffix>` flag, then the `test` field of `rustc_interface::Config` is true.

r? @camelid

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
In my opinion it's not necessarily the "right" usage. In tests you use items in order to make sure they work under any circumstance. Not sure if it's good to be used as an example.

**Comment by willcrichton** [comment]
> In my opinion it's not necessarily the "right" usage. In tests you use items in order to make sure they work under any circumstance. Not sure if it's good to be used as an example.

Note that this PR makes it *possible* for Rustdoc to scrape examples from tests. The proposed default is still that only examples and library code are used. But this way if someone wants to include tests as examples, then they can. That decision is left to the library author.

**Comment by GuillaumeGomez** [comment]
I stand on my position by if it's not enabled by default, then let's say I'm neutral.

**Comment by camelid** [comment]
I thought scrape-examples only works on `examples/`? Aren't `#[test]` functions usually in the crate itself?

**Comment by GuillaumeGomez** [comment]
They can be inside the crate source code and in their own folder.

**Comment by willcrichton** [comment]
Examples can theoretically be scraped from any source file in your Cargo workspace. The default is just the `examples/` directory.

**Comment by camelid** [comment]
Oh, so scrape-examples already supports scraping e.g. library source code if you tell it to?

**Comment by willcrichton** [comment]
That's right. And rust-lang/cargo#10343 makes that configurable at the target level.

**Comment by willcrichton** [comment]
If there are no other objections to this PR, it would be good to get it merged so I can get tests passing on rust-lang/cargo#10343.

**Comment by GuillaumeGomez** [comment]
This is quite an extension so I'd prefer to have everyone on the team agreeing with it first.

cc @rust-lang/rustdoc

**Comment by notriddle** [comment]
I support this PR. A crate might be intended for use in test cases, so it makes sense to have example code that uses `#[test]`, and you ought to be able up scrape it.

If other words, I think being example code and being test code are orthogonal.

**Comment by GuillaumeGomez** [comment]
I don't agree with your point of view but I agree on the fact that users should be able to include their tests into the "scraped examples". For me, tests are meant to check that your code *works* whereas examples are meant to show how to use your API.

**Comment by camelid** [comment]
Maybe this should be a different flag, e.g. `--include-tests`? `--test` is used to _run_ tests, but AFAICT this PR makes it so it's also used when building docs? That seems wrong.

**Comment by willcrichton** [comment]
This PR only affects the case where the user is scraping examples, not the general documentation paths. But I can create a new flag if you like.

**Comment by camelid** [comment]
> This PR only affects the case where the user is scraping examples, not the general documentation code paths. But I can create a new flag if you like.

Yeah, using the flag seems confusing to me.

**Comment by willcrichton** [comment]
Done.

**Comment by camelid** [comment]
Just noting that now that it's a new unstable flag, an FCP shouldn't be necessary.

**Comment by willcrichton** [comment]
I just added some documentation for the flag as well.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me now, thanks!

**Comment by willcrichton** [comment]
@GuillaumeGomez can you @bors (or should I do it?)

**Comment by GuillaumeGomez** [comment]
I'm waiting for one last review from @camelid. If they don't see anything else, then let's r+. :)

**Comment by camelid** [comment]
As long as this PR only adds behavior behind a new unstable flag, I have no objections.

**Comment by camelid** [comment]
#### Note

This PR's title is actually inaccurate since the contents changed after creation. It actually adds a new unstable flag and doesn't change the existing `--test` flag's behavior.

**Comment by GuillaumeGomez** [comment]
Good point. I updated the title.

**Pull request merged**
